runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/DataflowPipelineJob.java

```diff
@@ -263,19 +263,27 @@ State waitUntilFinish(
       }
 
       if (!hasError) {
-        // Reset the backoff.
+        // We can stop if the job is done.
+        if (state.isTerminal()) {
+          return state;
+        }
+
+        // The job is not done, so we must keep polling.
         backoff.reset();
-        // If duration is set, update the new cumulative sleep time to be the remaining
-        // part of the total input sleep duration.
+
+        // If a total duration for all backoff has been set, update the new cumulative sleep time to
+        // be the remaining total backoff duration, stopping if we have already exceeded the
+        // allotted time.
         if (duration.isLongerThan(Duration.ZERO)) {
           long nanosConsumed = nanoClock.nanoTime() - startNanos;
           Duration consumed = Duration.millis((nanosConsumed + 999999) / 1000000);
-          backoff =
-              MESSAGES_BACKOFF_FACTORY.withMaxCumulativeBackoff(duration.minus(consumed)).backoff();
-        }
-        // Check if the job is done.
-        if (state.isTerminal()) {
-          return state;
+          Duration remaining = duration.minus(consumed);
+          if (remaining.isLongerThan(Duration.ZERO)) {
+            backoff = MESSAGES_BACKOFF_FACTORY.withMaxCumulativeBackoff(remaining).backoff();
+          } else {
+            // If there is no time remaining, don't bother backing off.
+            backoff = BackOff.STOP_BACKOFF;
```

Comment:
nit: Why bother setting BackOff.STOP_BACKOFF? At this point, there is no time remaining. Why not just return the state? Or exit the loop?